addons/print/models/print_model.py

```diff
@@ -0,0 +1,295 @@
+        fields and method.
+    """
+
+    _name = 'print.provider'
+    _rec_name = 'name'
+    _description = 'Print Provider'
+
+    def _get_providers(self):
+        return []
+
+    # --------------------------------------------------
+    # MODEL FIELDS
+    # --------------------------------------------------
+    name = fields.Char("Name", required=True)
+    environment = fields.Selection([('test', 'Test'), ('production', 'Production')], "Environment", default='test')
+    provider = fields.Selection(selection='_get_providers', string='Provider', required=True)
+    balance = fields.Float("Credit", digits=(16, 2))
+
+    # --------------------------------------------------
+    # METHOD to be redefined in the provider implemenation
+    # --------------------------------------------------
+    @api.multi
+    def update_account_data(self):
+        """ Update the provider account data. Requires a fetch to the provider server. """
+        self.ensure_one()
+        if hasattr(self, '%s_update_account_data' % self.provider):
+            return getattr(self, '%s_update_account_data' % self.provider)()
+
+    def check_configuration(self):
+        """ Check if the credentials of the current provider are filled. If not, raise a warning. """
+        self.ensure_one()
+        if hasattr(self, '%s_check_configuration' % self.provider):
+            return getattr(self, '%s_check_configuration' % self.provider)()
+
+
+class print_order(models.Model):
+    """ Print Order Model. Each specific provider can extend the model by adding
+        its own fields, using the same convertion of the print.provider model.
+    """
+
+    _name = 'print.order'
+    _rec_name = 'id'
+    _description = 'Print Order'
+    _order = 'send_date desc'
+
+
+    def _default_print_provider(self):
+        return self.env['ir.values'].get_default('print.order', 'provider_id')
+
+    # --------------------------------------------------
+    # MODEL FIELDS
+    # --------------------------------------------------
+    create_date = fields.Datetime('Creation Date', readonly=True)
+    send_date = fields.Datetime('Sending Date', readonly=True)
+    currency_id = fields.Many2one('res.currency', 'Currency', required=True, default=lambda self: self.env.user.company_id.currency_id, readonly=True, states={'draft':[('readonly', False)]})
+    user_id = fields.Many2one('res.users', 'Author', default=lambda self: self.env.user)
+    provider_id = fields.Many2one('print.provider', 'Print Provider', required=True, default=_default_print_provider, readonly=True, states={'draft':[('readonly', False)]})
+
+    ink = fields.Selection([('BW', 'Black & White'), ('CL', 'Colour')], "Ink", default='BW', states={'sent':[('readonly', True)]})
+    paper_weight = fields.Integer("Paper Weight", default=80, readonly=True)
+    res_id = fields.Integer('Object ID')
+    res_model = fields.Char('Model Name')
```

Comment:
Can we have a print order not linked to a record ? As methods hereabove filter orders without model, maybe we could set those fileds as required and avoid having code filtering void res_modl.